"""
Snakefile for building the 2 trees of stdpopsim data, two population
"""
import sys
import numpy as np
import stdpopsim
sys.path.append("/code/src/methods/fsc26/")
import fastsimcoal2.
sys.path.append("/code/src/methods/dadi/")
import dadi_utils
import tskit
import pandas as pd
import plots

# ###############################################################################
# preperation :
# ###############################################################################

configfile: config["config"]+"/config.json"

np.random.seed(config["seed"])

# This is the number of samples to simulate for within each population
# for each replicate
# needs to be equal across sampled populations (or 0)
num_samples_per_population = config["num_samples_per_population"]
num_sampled_genomes_per_replicate = max(num_samples_per_population)

# The number of replicates of each analysis, # of seeds
replicates = config["replicates"]

# Where you would like all output files from analysis to live
output_dir = config["config"]

# The analysis species
species = stdpopsim.get_species(config["species"])

# The specific model you would like to run
model = species.get_demographic_model(config["model"])

# modifying the models events
generation_time = model.generation_time

mode_dict = {
  "reg": 140e3 / generation_time ,
  "high": 200e3 / generation_time
  }
# "low": 60e3 / generation_time


# identifier of chaned model use
model_modes = list(mode_dict.keys())

T_Div_loc = -1;
for i in range(len(model.demographic_events)):
    e = model.demographic_events[i]
    if e.type == "mass_migration":
        if e.source == 1 and e.dest == 0 and e.proportion == 1.0:
            T_Div_loc = i;

# The genetic map you would like to use.
# if value None is given default_recombination_rates are
# used with a flat map
genetic_map_id = config.get("genetic_map", None)

# The names of all chromosomes to simulate, separated by commas
# Use "all" to simulate all chromsomes for the genome
chrm_list = [chrom for chrom in species.genome.chromosomes]
if "chrY" in chrm_list:
    chrm_list.remove("chrY")
if(config["chrm_list"] != "all"):
    chrm_list = [chr for chr in config["chrm_list"].split(",")]
chrm = chrm_list[0]

# The model for inference with fsc
# Need different priors and mutation rates depending on species
fsc_model = "IM_hum"

# This grabs the default mr from the first chromosome,
# Ultimitely This needs to be replaced with the weighted average
# of all chromosomes: This should be done in stdpopsim.
mutation_rate = species.genome.mean_mutation_rate


# path for methods directory.
methods_dir_path = "/code/src/methods"

# ###############################################################################
# GENERAL RULES & GLOBALS
# ###############################################################################

seed_array = np.random.random_integers(1,2**31,replicates)

genetic_map_downloaded_flag= ".genetic_map_downloaded"

try:
    mask_file = config["mask_file"]
except KeyError:
    mask_file = None

rule all:
    input: expand([output_dir + "/{model_mode}/{seeds}/Results/estimates_N_tdiv_fsc.png", output_dir + "/{model_mode}/{seeds}/Results/estimates_mig_fsc.png"], model_mode=model_modes, seeds=seed_array)

rule download_genetic_map:
    output: genetic_map_downloaded_flag
    message: "Downloading default genetic map"
    run:
        # We need to have this here to avoid several threads trying to download the
        # the genetic map into the cache at the same time.
        if genetic_map_id is not None:
            genetic_map = species.get_genetic_map(genetic_map_id)
            if not genetic_map.is_cached():
                genetic_map.download()
            with open(output[0], "w") as f:
                print("File to indicate genetic map has been downloaded", file=f)


rule simulation:
    input:
        genetic_map_downloaded_flag
    output:
        output_dir + "/{model_mode}/{seeds}/Intermediate/" + chrm + ".trees"
    run:
        # use the right model based on the simulation run
        model.demographic_events[T_Div_loc].time = mode_dict[wildcards.model_mode]
        model.demographic_events[T_Div_loc + 1].time = mode_dict[wildcards.model_mode]
        engine = stdpopsim.get_default_engine()
        contig = species.get_contig(chrm, genetic_map=genetic_map_id)
        samples = model.get_samples(*num_samples_per_population)
        ts = engine.simulate(model, contig, samples, seed=wildcards.seeds)
        ts.dump(output[0])


# ###############################################################################
# dadi - prep for fsc
# ###############################################################################

rule ts_to_dadi_sfs:
    input:rules.simulation.output
    output: output_dir + "/{model_mode}/{seeds}/Intermediate/" + chrm + "_dadi_joint_sfs.fs",
            output_dir + "/{model_mode}/{seeds}/Intermediate/" + chrm + "_dadi_joint_sfs_with_nonvariant.fs"
    run:
        dadi_utils.ts_to_dadi_sfs(input[0], output[0],output[1], sample_size = num_sampled_genomes_per_replicate, mask_file=mask_file)

# ###############################################################################
# fastsimcoal2.
# ###############################################################################
 
rule dadi_sfs_to_fsc:
    input: rules.ts_to_dadi_sfs.output

    output: output_dir + "/{model_mode}/{seeds}/Intermediate/fsc_analysis/" + fsc_model + "_jointDAFpop1_0.obs",

    run:
        inputs = expand(output_dir + "/" + wildcards.model_mode + "/" + wildcards.seeds + "/Intermediate/" + chrm + "_dadi_joint_sfs_with_nonvariant.fs")
        dadi_out_path = output_dir + "/" + wildcards.model_mode + "/" + wildcards.seeds + "/Intermediate/all_chrm_dadi_joint_sfs_with_nonvariant.fs"
        fsc_out_path  = output_dir + "/" + wildcards.model_mode + "/" + wildcards.seeds + "/Intermediate/fsc_analysis/" + fsc_model + "_jointDAFpop1_0.obs"
        fastsimcoal2.dadi_to_fsc_sfs(inputs, dadi_out_path, fsc_out_path, sample_size=num_sampled_genomes_per_replicate)

rule fsc_setup:
    input: rules.dadi_sfs_to_fsc.output
    # expand(output_dir + "/{model_mode}/{seeds}/Intermediate/fsc_analysis/" + fsc_model + "_jointDAFpop1_0.obs")

    output: output_dir + "/{model_mode}/{seeds}/Intermediate/fsc_analysis/" + fsc_model + ".est",
            output_dir + "/{model_mode}/{seeds}/Intermediate/fsc_analysis/" + fsc_model + ".tpl"

    shell: "cp fsc_files/{fsc_model}.est {output_dir}/{wildcards.model_mode}/{wildcards.seeds}/Intermediate/fsc_analysis/{fsc_model}.est && \
            cp fsc_files/{fsc_model}.tpl {output_dir}/{wildcards.model_mode}/{wildcards.seeds}/Intermediate/fsc_analysis/{fsc_model}.tpl && \
            sed -i.bak 's/SAMPLE_SIZE/{num_sampled_genomes_per_replicate}/' {output_dir}/{wildcards.model_mode}/{wildcards.seeds}/Intermediate/fsc_analysis/{fsc_model}.tpl && rm {output_dir}/{wildcards.model_mode}/{wildcards.seeds}/Intermediate/fsc_analysis/{fsc_model}.tpl.bak &&\
            cd {methods_dir_path}/fsc26/ && chmod +x fsc26"

rule run_fsc:
    input: rules.fsc_setup.output[1]
    # expand(output_dir + "/{model_mode}/{seeds}/Intermediate/fsc_analysis/" + fsc_model + ".tpl")

    output: output_dir + "/{model_mode}/{seeds}/Intermediate/fsc_analysis/run10/" + fsc_model + "/" + fsc_model + ".bestlhoods"

    # I want to run this with a bash loop from 1..fit_runs to automate the number of fsc runs but for the life of me I cant get it to work in snakemake (it works fine in a shell script)
    # it keeps spitting out a python 'out of index' error and I've spent way too much time trying to troubleshoot it with no success
    # if someome can figure this out they would be my hero. Until that day, will continue to use this wall of bash code
    shell: "cd {output_dir}/{wildcards.model_mode}/{wildcards.seeds}/Intermediate/fsc_analysis/  && \
              mkdir run1 &&\
              cd run1 &&\
              cp ../{fsc_model}* ./ &&\
              {methods_dir_path}/fsc26/fsc26 fsc26 -t {fsc_model}.tpl -n 100000 -d -e {fsc_model}.est -M -L 40 -q -c 4 &&\
              cd .. &&\
              mkdir run2 &&\
              cd run2 &&\
              cp ../{fsc_model}* ./ &&\
              {methods_dir_path}/fsc26/fsc26 fsc26 -t {fsc_model}.tpl -n 100000 -d -e {fsc_model}.est -M -L 40 -q -c 4 &&\
              cd .. &&\
              mkdir run3 &&\
              cd run3 &&\
              cp ../{fsc_model}* ./ &&\
              {methods_dir_path}/fsc26/fsc26 fsc26 -t {fsc_model}.tpl -n 100000 -d -e {fsc_model}.est -M -L 40 -q -c 4 &&\
              cd .. &&\
              mkdir run4 &&\
              cd run4 &&\
              cp ../{fsc_model}* ./ &&\
              {methods_dir_path}/fsc26/fsc26 fsc26 -t {fsc_model}.tpl -n 100000 -d -e {fsc_model}.est -M -L 40 -q -c 4 &&\
              cd .. &&\
              mkdir run5 &&\
              cd run5 &&\
              cp ../{fsc_model}* ./ &&\
              {methods_dir_path}/fsc26/fsc26 fsc26 -t {fsc_model}.tpl -n 100000 -d -e {fsc_model}.est -M -L 40 -q -c 4 &&\
              cd .. &&\
              mkdir run6 &&\
              cd run6 &&\
              cp ../{fsc_model}* ./ &&\
              {methods_dir_path}/fsc26/fsc26 fsc26 -t {fsc_model}.tpl -n 100000 -d -e {fsc_model}.est -M -L 40 -q -c 4 &&\
              cd .. &&\
              mkdir run7 &&\
              cd run7 &&\
              cp ../{fsc_model}* ./ &&\
              {methods_dir_path}/fsc26/fsc26 fsc26 -t {fsc_model}.tpl -n 100000 -d -e {fsc_model}.est -M -L 40 -q -c 4 &&\
              cd .. &&\
              mkdir run8 &&\
              cd run8 &&\
              cp ../{fsc_model}* ./ &&\
              {methods_dir_path}/fsc26/fsc26 fsc26 -t {fsc_model}.tpl -n 100000 -d -e {fsc_model}.est -M -L 40 -q -c 4 &&\
              cd .. &&\
              mkdir run9 &&\
              cd run9 &&\
              cp ../{fsc_model}* ./ &&\
              {methods_dir_path}/fsc26/fsc26 fsc26 -t {fsc_model}.tpl -n 100000 -d -e {fsc_model}.est -M -L 40 -q -c 4 &&\
              cd .. &&\
              cd run10 &&\
              cp ../{fsc_model}* ./ &&\
              {methods_dir_path}/fsc26/fsc26 -t {fsc_model}.tpl -n 100000 -d -e {fsc_model}.est -M -L 40 -q -c 4"

# ###############################################################################
# Compile results and plot
# ###############################################################################

rule get_fsc_output:
    input: rules.run_fsc.output
    # expand(output_dir + "/{model_mode}/{seeds}/Intermediate/fsc_analysis/run10/" + fsc_model + "/" + fsc_model + ".bestlhoods")

    output: output_dir + "/{model_mode}/{seeds}/Intermediate/fsc_analysis/fsc_results_sorted.txt"
    
    run:
        path_to_fsc_analysis = output_dir+ "/" + wildcards.model_mode + "/" + wildcards.seeds + "/Intermediate/fsc_analysis/"
        fastsimcoal2.get_fsc_output(path_to_fsc_analysis, 10 ,"fsc_results_sorted.txt", fsc_model)
        
# see what it does
rule get_best_fsc_run:
    input: rules.get_fsc_output.output
    # expand(output_dir+"/{model_mode}/{seeds}/Intermediate/fsc_analysis/fsc_results_sorted.txt",model_mode=model_modes,seeds=seed_array)

    output: output_dir + "/{model_mode}/{seeds}/Results/best_fsc_run.txt"

    run:
        fastsimcoal2.get_best_fsc_run_general(path_to_fsc_analysis, output[0])   

rule plot_estimates:
    input: rules.get_best_fsc_run.output

    output: 
        output_dir + "/{model_mode}/{seeds}/Results/estimates_N_tdiv_fsc.png",
        output_dir + "/{model_mode}/{seeds}/Results/estimates_mig_fsc.png"

    run:

        simulated_genome_length = 0
        for chr in chrm_list:
            ts=tskit.load(output_dir + "/" + model_modes[0] + "/" + str(seed_array[0]) + "/Intermediate/" + chr + ".trees")
            simulated_genome_length += ts.sequence_length
            if mask_file:
                mask_table = pd.read_csv(mask_file, sep="\t", header=None)
                sub = mask_table[mask_table[0] == chr]
                mask_ints = pd.IntervalIndex.from_arrays(sub[1], sub[2])
                simulated_genome_length -= np.sum(mask_ints.length)
        T_Div = mode_dict[wildcards.model_mode]
        plots.plot_fsc_results_human_IM(input[0], output, simulated_genome_length, T_Div)

rule clean:
    shell:
        f"rm -rf {output_dir}/*/Intermediate \
            {output_dir}/*/Results \
            .genetic_map_downloaded \
            .snakemake"